### ISSUE TITLE
[Feature]: Record merkle root of withdrawal identifiers on chain

### ISSUE THREAD

**Issue by djordon** [stacks-network/sbtc#483]
# Feature - Record merkle root of withdrawal identifiers on the bitcoin blockchain

## 1. Description

For withdrawals, we do not have an airtight way of tying the withdrawal UTXO to the stacks transaction that initiated the request. When a signer tries to create a `withdrawal-accept` transaction it needs a knowing which UTXOs are associated with which `initiate-withdrawal-request` transactions. The original design for the withdrawal flow is in https://github.com/stacks-network/sbtc/issues/16.

There are two approaches that “work” to varying degrees without further changes to our design:

1. Have the signer who submitted the BTC transaction for the withdrawal create and submit the `withdrawal-accept` stacks transaction that includes this information.
2. Use the recipient, the amount, and maybe time to make an educated guess of the `initiate-withdrawal-request` transaction that is associated with a UTXO.

For (1), we cannot guarantee that signer who created the BTC transaction also creates the `withdrawal-accept` stacks transaction; and for (2) it would be better to have a way where we can “definitively” know which stacks transaction was associated with a UTXO. For deposits, we have a simple way of tying sweep transactions to deposits: the UTXO input in the sweep transaction always points to the deposit transaction.

### 1.1 Context & Purpose

The proposal here is three fold:

1. Change ordering of the UTXOs from the design in https://github.com/stacks-network/sbtc/issues/19: The first UTXO is the signers UTXO containing all deposits, the second UTXO will be the `OP_RETURN` UTXO, and the remaining UTXOs are for withdrawals. This is what is currently implemented in the code.
2. Change the OP byte in the `OP_RETURN` UTXO to a version byte. This version byte would allow us to change the layout of the data in the future by incrementing the version.
3. Include the root of a merkle tree at the end of the `OP_RETURN` UTXO described in https://github.com/stacks-network/sbtc/issues/19. 

The nodes of the merkle tree would be associated with the transaction’s UTXOs. Each node of the tree would be a hash160 of the withdrawal request’s identifier. This identifier is a concatenation of the stacks transaction ID that generated the withdrawal request, the request-id of the withdrawal request, and the stacks block ID that includes the transaction. We need the block Id since a transaction can be included in more than one block and we need the request-id because each transaction can generate more than one request. So the pre-image data for each node of the tree has the following layout:

```
0                 32                             64           72
|-----------------|------------------------------|------------|
  Stacks block ID   Txid of Stacks withdrawal tx   request-id
```

We represent the request-id as an 8 bytes big-endian unsigned integer. 

Including a merkle tree in the `OP_RETURN` would allow for a signer or outside observer to validate that particular withdrawal requests are included in the transaction. Also, including an extra 20 bytes for a merkle root is much cheaper than including the raw data that is necessary for the BTC transaction to point to all of the withdrawal requests in the bitcoin transaction.

After including the merkle root in the `OP_RETURN` UTXO, it will look like so:

```
0       2    3     5        21            41
|-------|----|-----|--------|-------------|
  magic   op   N_d   bitmap   merkle root
```

- `magic`: Two bytes that say that this is an sBTC transaction.
- `op`: the version of the data layout and it will start at `0u8`.
- `N_d`: The number of deposits in this transaction. These transactions can contain donations so it isn’t always obvious how many of the inputs were deposits (you’d have to inspect the script pubkey and maybe the witness data).
- `signer_bitmap`: Bitmap of signer votes. This is only included if the transaction includes deposit inputs and/or withdrawal outputs. If the transaction only has donations then this will be missing.
- `merkle root`: The root of the merkle tree describing the withdrawal UTXOs. This is only included if the transaction includes withdrawal outputs. If the transaction only has donations or deposits then this will be missing.

This is mostly the same layout as in https://github.com/stacks-network/sbtc/issues/19, except that the meaning of the `op` has changed and the merkle root has been added at the end.

### Prior proposal

Note: This was the proposal here before https://github.com/stacks-network/sbtc/issues/483#issuecomment-2333252631 and it doesn’t work. It requires bitcoin transactions with multiple `OP_RETURN` UTXOs, which is non-standard and so won’t be relayed by bitcoin-core nodes.

The proposal is to have each withdrawal UTXO have an associated `OP_RETURN` UTXO that contains the txid of the stacks withdrawal request transaction that lead to the withdrawal UTXO as well as the associated request-id that was generated by the stacks transaction. The data layout would be:

```
0         1                              33           37
|---------|------------------------------|------------|
  version   txid of stacks withdrawal tx   request-id

```

The version byte is in case we want to update the data layout in the future, and will start out as `0u8`. A single stacks transaction can be associated with more than one withdrawal request, so the request-id is used to disambiguate them within a transaction. Also note that because of reorgs, a request-id need not point to a unique stacks transaction, so we need both the txid and the request-id. And lastly, we represent the request-id as a 4 bytes big-endian unsigned integer, instead of 8 or 16 bytes since 4 bytes will last at least 130 years (assuming no more than one withdrawal request per second on average).

These `OP_RETURN` outputs would follow immediately after their associated withdrawal UTXO.

(Alternatively, they could all follow after all of the withdrawal UTXOs but in the same order of their associated withdrawal UTXO.)

The downside of this proposal is that it makes withdrawals more expensive. Each withdrawal now weighs an additional 48 vbytes, which can be pricey. I came up with this value by running the following:

```rust
#[test]
fn testme() {
    use bitcoin::TxOut;
    use bitcoin::Amount;
    use bitcoin::ScriptBuf;

    let tx_out = TxOut {
        value: Amount::ZERO,
        script_pubkey: ScriptBuf::new_op_return([0; 37]),
    };
    println!("{}", tx_out.weight().to_vbytes_ceil());
}

```

### 2.1 Acceptance Criteria:

- [ ]  Add `OP_RETURN` UTXOs to our sweep transactions servicing withdrawal requests.

## 3. Related Issues and Pull Requests (optional):

- The withdrawal flow design https://github.com/stacks-network/sbtc/issues/16
- The design for the BTC transaction layout https://github.com/stacks-network/sbtc/issues/19

**Comment by netrome** [comment]
For reference, the sBTC transaction format is defined in https://github.com/stacks-network/sbtc/issues/19

**Comment by netrome** [comment]
I have contemplated a bit on this, and I think the general idea makes sense.

In terms of storage space, I suppose we only need to tie the withdraw UTXOs to the corresponding `request_id`, which is 16 bytes, so that saves us almost half the space compared to using the transaction. If this is a problem we could consider using a more compact encoding scheme, as the request IDs are incremental and start at 0 afaik.

**Comment by netrome** [comment]
One thing to consider here is that the clarity bitcoin library only sees a limited number of UTXOs in a single bitcoin transaction, so a variant where this information is concatenated into a single OP_RETURN output would be preferable - though this adds additional limitations in how many withdraw requests we can process.

**Comment by djordon** [comment]
> One thing to consider here is that the clarity bitcoin library only sees a limited number of UTXOs in a single bitcoin transaction, so a variant where this information is concatenated into a single OP_RETURN output would be preferable - though this adds additional limitations in how many withdraw requests we can process.

Ohh right, I had forgotten about the clarity bitcoin limitation. Do you know more about the nature of this limitation?

**Comment by netrome** [comment]
> Ohh right, I had forgotten about the clarity bitcoin limitation. Do you know more about the nature of this limitation?

All I know is that clarity can only see the first 8 inputs and outputs of a transaction, as documented in #19. This stems from the clarity-bitcoin contract, and I'd assume it being possible to update the contract to use any other number. I think the 8 is an arbitrary value here, and the real limit is that it cannot be unbounded due to the turing-incomplete nature of Clarity.

![image](https://github.com/user-attachments/assets/7c79787e-ba5a-4029-a318-fc13ea6c3fe1)

**Comment by cylewitruk** [comment]
Wrote this on Slack but Daniel asked me to post it here too:

One alternative would be to embrace the P2P network and use alt. 1, but broadcasting intent and confirmation on the P2P network before and after the bitcoin transaction. This would allow for another signer to pick up the slack if the bitcoin transaction never makes it to the bitcoin mempool. The information could either be stored in local DB's or in the network DHT until the bitcoin transaction is confirmed.

This approach also allows the signer network to monitor the bitcoin transaction for inclusion in a block, RBF if needed, etc. even in the event the "owning signer" would go offline.

Daniel mentioned that the original plan was for the "owning signer" to simply store the information in its local DB, but this doesn't protect against signer failure between persistence and broadcast.

**Comment by djordon** [comment]
I just realized that a single stacks transaction ID might have more than one associated withdrawal, so if we write the transaction ID on chain, we probably want to include the `request-id` too. If we write both the txid and the request-id on chain that means the size of the `OP_RETURN` UTXO weights 51 bytes.

**Comment by djordon** [comment]
> I have contemplated a bit on this, and I think the general idea makes sense.
> 
> In terms of storage space, I suppose we only need to tie the withdraw UTXOs to the corresponding `request_id`, which is 16 bytes, so that saves us almost half the space compared to using the transaction. If this is a problem we could consider using a more compact encoding scheme, as the request IDs are incremental and start at 0 afaik.
> 
> EDIT: Note that using request ID rather than txid can become problematic if we have a deep bitcoin fork, which could potentially alter the transaction order on Stacks. So sticking with txid here may be the safer option.

If we have a deep bitcoin reorg that alters the transaction that made the `initiate-withdrawal-request` contract call, then we have other issues right? We wait before acting on these requests so that we do not have to worry about bitcoin reorgs anyway. So I think we can just use the `request-id` if we want to. Moreover, I'd just shrink the `request-id` to 8 bytes, we'd have other issues before we exhaust that number of bytes.

**Comment by djordon** [comment]
> If we have a deep bitcoin reorg that alters the transaction that made the `initiate-withdrawal-request` contract call, then we have other issues right? We wait before acting on these requests so that we do not have to worry about bitcoin reorgs anyway. So I think we can just use the `request-id` if we want to. Moreover, I'd just shrink the `request-id` to 8 bytes, we'd have other issues before we exhaust that number of bytes.

Wait, there can be ambiguity when using `request-id`s alone, since the reorg doesn't need to be deep in order to change them. This means you cannot have a `request-id` be the sole identifier, which I now realize was your point 😅.

**Comment by netrome** [comment]
> If we have a deep bitcoin reorg that alters the transaction that made the initiate-withdrawal-request contract call, then we have other issues right?

Yeah, because it may look like the signers have stolen funds without minting anything on stacks :sweat_smile:

**Comment by djordon** [comment]
Ugh, this approach cannot work either, since there can only be one `OP_RETURN` UTXO per transaction. I don't understand why, maybe to discourage their use.

**Comment by cylewitruk** [comment]
> Ugh, this approach cannot work either, since there can only be one `OP_RETURN` UTXO per transaction. I don't understand why, maybe to discourage their use.

Thinking on @netrome's comment above re: storing everything in one `OP_RETURN` -- does `OP_RETURN` really need to be used? I read that there's no validation being done on the Clarity side for v1 (right?).

Could we take some inspiration from ordinal inscriptions and store the data in a taproot spend script which gives 4MB of storage at the discounted "witness rate"? I.e. something like:
```
OP_FALSE
OP_IF
  OP_PUSH "sbtc"
  OP_PUSH 2
  OP_PUSH <TX1_DATA>
  OP_PUSH <TX2_DATA>
OP_ENDIF
```

Just a random thought, no idea if it would work for this case in reality, but that's one way others are storing data on Bitcoin. Note it requires commit/reveal. If the Clarity-side wants to do validation then we could probably add support in Clarity post-nakamoto.

**Comment by djordon** [comment]
Great suggestions, let me try to respond to everything here.

> Thinking on @netrome's comment above re: storing everything in one `OP_RETURN` -- does `OP_RETURN` really need to be used?

It doesn't need to be an `OP_RETURN` UTXO, but it's the cheapest. And increasing the cost even more caused me to re-evaluate whether this is really the way. Also, for non-`OP_RETURN` UTXOS we should think more carefully about who can spend them and what to actually do.

> I read that there's no validation being done on the Clarity side for v1 (right?).

That's right, there is no validation done on the clarity side.

> I could be way off base here, but would it be possible to take some inspiration from ordinal inscriptions and store the data in a taproot spend script which gives 4MB of storage at 1/4 the cost (witness discount)?
>
> Just a random thought, no idea if it would work for this case in reality, but that's one way others are storing data on on-chain. Note it requires commit/reveal. 

Not off base at all, these are good ideas. There are other similar designs that we can consider (like maybe we include a merkle tree with this data in the existing `OP_RETURN`). But yeah it takes a little more thought to figure out what would be "best".

**Comment by djordon** [comment]
Okay, I updated the proposal in the opening description. 

TLDR: include the root of a merkle tree in our existing `OP_RETURN` UTXO. The pre-image of the leaf nodes in the tree are the identifiers of the withdrawal requests that are included in the transaction.

This allows for a more simple (but tedious) validation by the signers and outside observers. The proposal also costs just 20 bytes on chain.
